src/vmspawn/vmspawn.c

```diff
@@ -790,6 +819,19 @@ static int run_virtual_machine(void) {
                 r = strv_extend_many(&cmdline, "-kernel", arg_linux);
                 if (r < 0)
                         return log_oom();
+
+                /* We can't rely on gpt-auto-generator when direct kernel booting so synthesize a root=
+                 * kernel argument instead. */
+                if (arg_image && !strv_find_startswith(arg_kernel_cmdline_extra, "root=")) {
+                        _cleanup_free_ char *root = NULL;
+                        r = finalize_root(&root);
+                        if (r < 0)
+                                return log_error_errno(r, "Cannot perform a direct kernel boot without a root or usr partition: %m");
+
+                        r = strv_extend(&arg_kernel_cmdline_extra, root);
+                        if (r < 0)
+                                return log_oom();
+                }
```

Comment:
i'd move this whole if block into a helper func `kernel_cmdline_maybe_append_root()` or something?